File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/AstGeneratorResult2RichResultConverter.java
package titan.ast.runtime;

import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import titan.ast.runtime.AstGeneratorResult.AstParseErrorData;
import titan.ast.runtime.AstGeneratorResult.TokenParseErrorData;
import titan.ast.runtime.AstGeneratorResult.TokensResult;
import titan.ast.runtime.LineNumberDetail.LineNumberRange;
import titan.ast.runtime.RichAstGeneratorResult.RichAstParseErrorData;
import titan.ast.runtime.RichAstGeneratorResult.RichAstResult;
import titan.ast.runtime.RichAstGeneratorResult.RichTokenParseErrorData;
import titan.ast.runtime.RichAstGeneratorResult.RichTokensResult;

/**
 * .
 *
 * @author tian wei jun
 */
class AstGeneratorResult2RichResultConverter {
  RichAstGeneratorResultStringEncoder stringEncoder = new RichAstGeneratorResultStringEncoder();
  private byte newline = '\n';

  byte getNewline() {
    return newline;
  }

  void setNewline(byte newline) {
    this.newline = newline;
  }

  Charset getCharset() {
    return stringEncoder.getCharset();
  }

  void setCharset(String charsetName) {
    stringEncoder.setCharset(charsetName);
  }

  void setCharset(Charset charset) {
    stringEncoder.setCharset(charset);
  }

  RichAstGeneratorResult convert(AstGeneratorResult astGeneratorResult) {
    RichTokensResult richTokensResult =
        convert2RichTokensResultNoEncodingTokens(astGeneratorResult.tokensResult);
    RichAstResult richAstResult = convert2RichAstResultNoEncodingAst(astGeneratorResult);
    encode(richTokensResult, richAstResult);
    return new RichAstGeneratorResult(richTokensResult, richAstResult);
  }

  private void encode(RichTokensResult richTokensResult, RichAstResult richAstResult) {
    switch (richAstResult.getType()) {
      case OK -> {
        ArrayList<Token> tokens = richTokensResult.getOkData();
        stringEncoder.encodeTokens(tokens);
        encodeAstByEncodedTokens(richAstResult.getOkData(), tokens.iterator());
      }
      case AST_PARSE_ERROR -> {
        stringEncoder.encodeTokens(richTokensResult.getOkData());
      }
      case TOKENS_ERROR -> {}
    }
  }

  private void encodeAstByEncodedTokens(Ast ast, Iterator<Token> encodedTokensIt) {
    for (Ast child : ast.children) {
      encodeAstByEncodedTokens(child, encodedTokensIt);
    }
    if (ast instanceof TerminalAst terminalAst) {
      Token encodedToken = encodedTokensIt.next();
      while (encodedToken.type != TokenType.TEXT) {
        encodedToken = encodedTokensIt.next();
      }
      terminalAst.token.text = encodedToken.text;
      terminalAst.token.start = encodedToken.start;
    }
  }

  private RichAstResult convert2RichAstResultNoEncodingAst(AstGeneratorResult astGeneratorResult) {
    RichAstResult richAstResult = null;

    switch (astGeneratorResult.astResult.getType()) {
      case OK -> {
        richAstResult = RichAstResult.generateOkResult(astGeneratorResult.astResult.getOkData());
      }
      case AST_PARSE_ERROR -> {
        richAstResult =
            RichAstResult.generateRichAstParseErrorResult(
                convert2RichAstParseErrorData(
                    astGeneratorResult.astResult.getAstParseErrorData(),
                    astGeneratorResult.tokensResult.getOkData()));
      }
      case TOKENS_ERROR -> {
        richAstResult = RichAstResult.generateRichTokensErrorResult();
      }
    }
    return richAstResult;
  }

  private RichAstParseErrorData convert2RichAstParseErrorData(
      AstParseErrorData astParseErrorData, ArrayList<Token> tokens) {
    LineNumberDetail lineNumberDetail = buildLineNumberDetail(tokens);
    LineNumberRange startLineNumberRange =
        lineNumberDetail.getLineNumberRange(astParseErrorData.start);
    LineNumberRange endLineNumberRange =
        lineNumberDetail.getLineNumberRange(astParseErrorData.end - 1);
    int startOffsetInLine =
        stringEncoder.getOffsetInLine(tokens, startLineNumberRange, astParseErrorData.start);
    int endOffsetInLine =
        stringEncoder.getOffsetInLine(tokens, endLineNumberRange, astParseErrorData.end);
    return new RichAstParseErrorData(
        startLineNumberRange.lineNumber,
        startOffsetInLine + 1, // 从1开始计数，所以+1
        endLineNumberRange.lineNumber,
        endOffsetInLine + 1, // 从1开始计数，所以+1
        stringEncoder.encodeString(astParseErrorData.errorText));
  }

  private RichTokenParseErrorData convert2RichTokenGeneratorErrorData(
      TokenParseErrorData tokenParseErrorData) {
    ArrayList<Token> tokens = new ArrayList<>(tokenParseErrorData.finishedTokens.size() + 1);
    tokens.addAll(tokenParseErrorData.finishedTokens);
    Token errorToken = new Token(tokenParseErrorData.start);
    errorToken.text = tokenParseErrorData.errorText;
    tokens.add(errorToken);

    LineNumberDetail lineNumberDetail = buildLineNumberDetail(tokens);

    LineNumberRange startLineNumberRange =
        lineNumberDetail.getLineNumberRange(tokenParseErrorData.start);
    LineNumberRange endLineNumberRange =
        lineNumberDetail.getLineNumberRange(tokenParseErrorData.end - 1);
    int startOffsetInLine =
        stringEncoder.getOffsetInLine(tokens, startLineNumberRange, tokenParseErrorData.start);
    int endOffsetInLine =
        stringEncoder.getOffsetInLine(tokens, endLineNumberRange, tokenParseErrorData.end);

    return new RichTokenParseErrorData(
        stringEncoder.encodeTokens(tokenParseErrorData.finishedTokens),
        startLineNumberRange.lineNumber,
        startOffsetInLine + 1, // 从1开始计数，所以+1
        endLineNumberRange.lineNumber,
        endOffsetInLine + 1, // 从1开始计数，所以+1
        stringEncoder.encodeString(errorToken.text));
  }

  private RichTokensResult convert2RichTokensResultNoEncodingTokens(TokensResult tokensResult) {
    RichTokensResult richTokensResult = null;
    switch (tokensResult.getType()) {
      case OK -> {
        richTokensResult = RichTokensResult.generateOkResult(tokensResult.getOkData());
      }
      case TOKEN_PARSE_ERROR -> {
        richTokensResult =
            RichTokensResult.generateTokenParseErrorResult(
                convert2RichTokenGeneratorErrorData(tokensResult.getTokenParseErrorData()));
      }
      case SOURCE_IO_ERROR -> {
        richTokensResult =
            RichTokensResult.generateSourceIoErrorResult(tokensResult.getSourceIoErrorData());
      }
    }
    return richTokensResult;
  }

  private LineNumberDetail buildLineNumberDetail(List<Token> tokens) {
    ArrayList<LineNumberRange> lineNumberRanges = new ArrayList<LineNumberRange>();
    int indexOfStartToken = 0;
    int start = 0;
    int indexOfBytes = -1;
    int indexOfToken = 0;
    for (; indexOfToken < tokens.size(); indexOfToken++) {
      Token token = tokens.get(indexOfToken);
      byte[] bytes = token.text.getBytes(AstGeneratorResult.DEFAULT_CHARSET);
      for (int indexOfTokenBytes = 0; indexOfTokenBytes < bytes.length; indexOfTokenBytes++) {
        ++indexOfBytes;
        if (bytes[indexOfTokenBytes] == newline) {
          int end = indexOfBytes + 1;
          int indexOfEndToken = indexOfToken;
          lineNumberRanges.add(
              new LineNumberRange(
                  start, end, lineNumberRanges.size() + 1, indexOfStartToken, indexOfEndToken));
          // 更新下一行
          start = end;
          if (indexOfTokenBytes + 1 < bytes.length) {
            indexOfStartToken = indexOfEndToken;
          } else {
            indexOfStartToken = indexOfEndToken + 1;
          }
        }
      }
    }
    // 最后一行可以没有换行符，设置这个特殊行
    if (indexOfBytes >= start) {
      int end = indexOfBytes + 1;
      lineNumberRanges.add(
          new LineNumberRange(
              start, end, lineNumberRanges.size() + 1, indexOfStartToken, indexOfToken));
      // 更新下一行
      start = end;
    }
    // lineNumberRanges move to array
    LineNumberRange[] lineNumberRangeArray = new LineNumberRange[lineNumberRanges.size()];
    int indexOfLineNumberRange = 0;
    for (LineNumberRange lineNumberRange : lineNumberRanges) {
      lineNumberRangeArray[indexOfLineNumberRange++] = lineNumberRange;
    }
    return new LineNumberDetail(lineNumberRangeArray);
  }
}

File: runtime/cpp/titan-ast-runtime-lib/AstGeneratorResult2RichResultConverter.cpp
//
// Created by tian wei jun on 2024/10/26.
//

#include "AstGeneratorResult2RichResultConverter.h"
AstGeneratorResult2RichResultConverter::AstGeneratorResult2RichResultConverter()
    : newline('\n') {}
AstGeneratorResult2RichResultConverter::
    ~AstGeneratorResult2RichResultConverter() = default;

void AstGeneratorResult2RichResultConverter::setNewline(byte newline) {
  this->newline = newline;
}
byte AstGeneratorResult2RichResultConverter::getNewline() { return newline; }

RichAstGeneratorResult *AstGeneratorResult2RichResultConverter::convert(
    AstGeneratorResult *astGeneratorResult) {
  RichAstResult *richAstResult = convert2RichAstResult(astGeneratorResult);
  RichTokensResult *richTokensResult =
      convert2RichTokensResult(astGeneratorResult->tokensResult);
  delete astGeneratorResult;
  return new RichAstGeneratorResult(richTokensResult, richAstResult);
}

RichTokensResult *
AstGeneratorResult2RichResultConverter::convert2RichTokensResult(
    TokensResult *tokensResult) {
  RichTokensResult *richTokensResult = nullptr;
  switch (tokensResult->type) {
  case TokensResultType::OK: {
    richTokensResult =
        RichTokensResult::generateOkResult(tokensResult->getOkData());
    // data move
    break;
  }
  case TokensResultType::TOKEN_PARSE_ERROR: {
    auto tokenParseErrorData = tokensResult->getTokenParseErrorData();
    richTokensResult = RichTokensResult::generateTokenParseErrorResult(
        convert2RichTokenGeneratorErrorData(tokenParseErrorData));
    // data delete
    delete tokenParseErrorData;
    break;
  }
  case TokensResultType::SOURCE_IO_ERROR: {
    richTokensResult = RichTokensResult::generateSourceIoErrorResult(
        tokensResult->getSourceIoErrorData());
    // data move
    break;
  }
  }
  // data is invalid
  tokensResult->data = nullptr;
  return richTokensResult;
}

LineNumberDetail *AstGeneratorResult2RichResultConverter::buildLineNumberDetail(
    std::vector<Token *> *tokens) {
  char charNewline = (char)newline;
  std::list<LineNumberRange> lineNumberRanges;
  int nextStart = 0;
  int indexOfBytes = -1;
  for (Token *token : *tokens) {
    auto text = token->text.data();
    auto lenOfText = token->text.length();
    for (int indexOfText = 0; indexOfText < lenOfText; indexOfText++) {
      char ch = text[indexOfText];
      ++indexOfBytes;
      if (ch == charNewline) {
        int nextEnd = indexOfBytes + 1;
        lineNumberRanges.push_back(LineNumberRange(nextStart, nextEnd));
        // 更新下一行
        nextStart = nextEnd;
      }
    }
  }
  // 最后一行可以没有换行符，设置这个特殊行
  if (indexOfBytes >= nextStart) {
    int nextEnd = indexOfBytes + 1;
    lineNumberRanges.push_back(LineNumberRange(nextStart, nextEnd));
    // 更新下一行
    nextStart = nextEnd;
  }
  // lineNumberRanges move to array
  auto *lineNumberRangeArray = new LineNumberRange[lineNumberRanges.size()];
  int indexOfLineNumberRange = 0;
  for (const LineNumberRange &lineNumberRange : lineNumberRanges) {
    lineNumberRangeArray[indexOfLineNumberRange++] = lineNumberRange;
  }
  return new LineNumberDetail(lineNumberRangeArray, lineNumberRanges.size());
}
RichAstResult *AstGeneratorResult2RichResultConverter::convert2RichAstResult(
    AstGeneratorResult *astGeneratorResult) {
  RichAstResult *richAstResult = nullptr;
  switch (astGeneratorResult->astResult->type) {
  case AstResultType::OK: {
    richAstResult = RichAstResult::generateOkResult(
        astGeneratorResult->astResult->getOkData());
    // data move
    break;
  }
  case AstResultType::AST_PARSE_ERROR: {
    auto astParseErrorData =
        astGeneratorResult->astResult->getAstParseErrorData();
    richAstResult = RichAstResult::generateRichAstParseErrorResult(
        convert2RichAstParseErrorData(
            astGeneratorResult->astResult->getAstParseErrorData(),
            astGeneratorResult->tokensResult->getOkData()));
    // delete data
    delete astParseErrorData;
    break;
  }
  case AstResultType::TOKENS_ERROR: {
    richAstResult = RichAstResult::generateRichTokensErrorResult();
    // data is nullptr
    break;
  }
  }
  // data is invalid,been moved
  astGeneratorResult->astResult->data = nullptr;
  return richAstResult;
}

RichTokenParseErrorData *
AstGeneratorResult2RichResultConverter::convert2RichTokenGeneratorErrorData(
    TokenParseErrorData *tokenParseErrorData) {
  std::vector<Token *> tokens(tokenParseErrorData->finishedTokens->size() + 1,
                              nullptr);
  int indexOfToken = 0;
  for (auto token : *tokenParseErrorData->finishedTokens) {
    tokens[indexOfToken++] = token;
  }
  Token errorToken;
  errorToken.start = tokenParseErrorData->start;
  errorToken.text = tokenParseErrorData->errorText;
  tokens[indexOfToken++] = &errorToken;

  LineNumberDetail *lineNumberDetail = buildLineNumberDetail(&tokens);
  LineNumberRangeDto startLineNumberRange =
      lineNumberDetail->getLineNumberRangeDto(tokenParseErrorData->start);
  LineNumberRangeDto endLineNumberRange =
      lineNumberDetail->getLineNumberRangeDto(tokenParseErrorData->end - 1);
  delete lineNumberDetail;
  auto richTokenParseErrorData = new RichTokenParseErrorData(
      tokenParseErrorData->finishedTokens, tokenParseErrorData->start,
      tokenParseErrorData->end, startLineNumberRange.lineNumber,
      tokenParseErrorData->start - startLineNumberRange.start +
          1, // 用户角度下标从1开始
      endLineNumberRange.lineNumber,
      tokenParseErrorData->end - endLineNumberRange.start +
          1, // 用户角度下标从1开始
      tokenParseErrorData->errorText);
  // take token in finishedTokens
  tokenParseErrorData->finishedTokens = nullptr;
  return richTokenParseErrorData;
}

RichAstParseErrorData *
AstGeneratorResult2RichResultConverter::convert2RichAstParseErrorData(
    AstParseErrorData *astParseErrorData, std::vector<Token *> *tokens) {
  LineNumberDetail *lineNumberDetail = buildLineNumberDetail(tokens);
  LineNumberRangeDto startLineNumberRange =
      lineNumberDetail->getLineNumberRangeDto(astParseErrorData->start);
  LineNumberRangeDto endLineNumberRange =
      lineNumberDetail->getLineNumberRangeDto(astParseErrorData->end - 1);
  delete lineNumberDetail;
  return new RichAstParseErrorData(
      astParseErrorData->start, astParseErrorData->end,
      startLineNumberRange.lineNumber,
      astParseErrorData->start - startLineNumberRange.start +
          1, // 用户角度下标从1开始
      endLineNumberRange.lineNumber,
      astParseErrorData->end - endLineNumberRange.start +
          1, // 用户角度下标从1开始
      astParseErrorData->errorText);
}
